Fix unreliable entities pushed by pistons

petersv5 opened the pull request:
[resubmission of a corrected pull as I accidentally closed the previous #6301]

This patch solves #5189 where entities in slime blocks are unreliable
in PaperMC. This is known to break some contraptions like the ice road
maker mentionend there.

The removal of 1/4 of all ticks for an entity that takes place in
ActivationRange.checkIfActive() results in an entity like e.g.
an armor stand not noticing that it has had setDeltaMovement() called.
The result is that an armor stand inside a slime block gets left behind
1/4 of the pushes.

There were already some hard coded logic in checkIfActive to not apply
the culling in some circumstances. The attach patch takes a slightly
different approach and adds a general mechanism where the entity knows
how many ticks into the future that it has to be fully ticked and
ActivationRange queries that.

Another approach would be to just have a "wasMovedByPiston" memory in
the entity and have the timing logic in ActivationRange. I prepared a
patch for that as well in case that is preferable. All things being
equal I thought it better to submit the more general patch if more cases
of needing to temporarily override the tick culling.

As dicussed on the discord channel this patch does not affect the
configration option to not tick armor stands as that is handled
in ArmorStand.tick().

Proximyst commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,76 @@
+ 
+         this.firstTick = false;
+         this.level.getProfiler().pop();
++        // Paper start   - track when all ticks are required for the entity, e.g. after a piston push
```
What's with the extra spaces all over the place?

petersv5 replied:
Which extra spaces are you referring to? The ones inside the //Paper comments or are there others I should fix? 
Should I fix them on my side?

Proximyst replied:
Yeah, the ones inside the paper comments. If you feel like fixing them, sure, otherwise we'll take care of it on merge (though it'd be kindest if you did 😄).

petersv5 replied:
Updated in the new commit

Proximyst replied:
You have several instances of this :)

petersv5 replied:
I fixed the two instances of extra spaces I found. Were there any more?

Proximyst commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,76 @@
+     // Paper end
++ 
++    // Paper start   - track when all ticks are required for the entity, e.g. after a piston push
++    public int remainingExactTicks = 0;
```
I'd prefer we make this private with a getter; `requestExactTicks` already acts as a setter, so might as well do proper encapsulation.

petersv5 replied:
Ok, not a problem. I started by adding getter/setters, then pulled those out following a pattern I saw in other places. Easy enough to put back.

petersv5 replied:
Done in the new commit. Let me know if you perfer a different naming convention.

Proximyst commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,76 @@
+     private static void moveEntityByPiston(Direction direction, Entity entity, double d, Direction direction2) {
+         NOCLIP.set(direction);
+         entity.move(MoverType.PISTON, new Vec3(d * (double)direction2.getStepX(), d * (double)direction2.getStepY(), d * (double)direction2.getStepZ()));
++        entity.requestExactTicks(10);    // Paper - track when all ticks are required for the entity, e.g. after a piston push
```
Have you tested with lower values?

petersv5 replied:
I belive that the moveCollidedEntities is also called for "bounced", and I thought it best to give enough time for the movement to die down. For the stuck entities 2 ticks should be enough. I'll test that, as well if bounced entities are affected by this patch. Back in 10 minutes.

Proximyst replied:
If you have some source other places where you pulled the number, feel free to keep it, but add a comment on the line that has said number.

petersv5 replied:
After a bit of testing: for items stuck inside a slime block 2 ticks is sufficient to guarantee it starts and stops moving. 
This function is also called for non-stuck items, that are laucnhed ballistically. I think they will move differently depending on if they get all ticks or not while in motion. Only 2 ticks are needed for the ice maker and similar machines based on stuck entities.

On the other hand, I think paper already accepts non-vanilla behaviour for dynamic movement as the 1/4 tick removal exists, so maybe we should lower tihs value to 2? We can always make it configurable in the future if good test cases are found. I suspect that perhaps tnt ender pearl cannons would be impacted by this, but I have no good test case for this.

petersv5 replied:
Updated with a chnage to 2 ticks only. This was verfied as sufficient both manually and with the Ilmango deisgn from issue 5189 running over 1200 blocks without issue. Bounced item trajectories are leaft for the future, once someone designs a test case for comparing differences between vanilla and paper.

Proximyst replied:
I suppose a test for some ender pearl device could be done to make sure they remain functional? Otherwise, I'm sure the various broken machines will turn up in the issue tracker after merge, anyway.

petersv5 replied:
Any machines working before the fix should see no change after the fix, excpet perhaps a tiny bit less variation. It is likely still not vanilla behaviour for bounced entities, but I would need a test case before we delve into that. I suggest we leave that for a future issue...

Proximyst reviewed (changes requested):
Blocking on answer for lower values; the rest could be handled pretty easily by the merger.

Proximyst commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
+         this.firstTick = false;
+         this.level.getProfiler().pop();
++        // Paper start - track when all ticks are required for the entity, e.g. after a piston push
++        if (remainingExactTicks > 0) {
```
Might as well use `this.isTickRequired` here.

petersv5 replied:
Done in new commit.

Proximyst commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
+         this.firstTick = false;
+         this.level.getProfiler().pop();
++        // Paper start - track when all ticks are required for the entity, e.g. after a piston push
++        if (remainingExactTicks > 0) {
```
Also prefer using `this.` prefixes; not at all a blocker, however.

petersv5 replied:
Done in new commit.

Proximyst commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
+                                 }
+ 
+                                 entity.setDeltaMovement(e, g, h);
++                                entity.requestExactTicks(2);    // Paper - track when all ticks are required for the entity, e.g. after a piston push
```
This and the below hunk need reformatting

petersv5 replied:
Will you handle that as a part of the merge, or do you want me to fix it and resubmit?

Proximyst replied:
Either or; your approval only gets removed if a larger change comes

petersv5 replied:
I'll be away for the rest of the day.  If it is not already merged when I get back I can update the formatting.

petersv5 replied:
I am not sure what formatting is requested. Should I switch to the multi-line format for the // Paper marker, or was it something else you had in mind?

qixils replied:
i think she means the 4 spaces before the comment. everything else in paper just uses 1 (`; // Paper - `)

petersv5 replied:
Updated. let me know if I got this right.

NoahvdAa commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
++    public boolean isTickRequired() {
++        return (this.remainingExactTicks > 0);
++    }
++    // Peter end
```
```suggestion
+    // Paper end
```

NoahvdAa commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
+         if (entity instanceof Mob && ((Mob)entity).leashHolder instanceof Player) {
+             return true;
+         }
++        // immunize entities requesting exact ticks, e.g. being affected by a piston
```
```suggestion
+        // Paper start - immunize entities requesting exact ticks, e.g. being affected by a piston
```

qixils replied:
unnecessary, this is part of an existing paper block

NoahvdAa replied:
Whoops, didn't see that. Is there a way for me to mark this as solved, or is that something only @petersv5 can do?

qixils commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
++    public boolean isTickRequired() {
++        return (this.remainingExactTicks > 0);
++    }
++    // Peter end
```
since there's already a paper block above this, you can just move this into the other paper block (i.e. remove lines 32-34)

petersv5 replied:
Updated in new commit

NoahvdAa reviewed (approved):
LGTM!

Machine-Maker commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
+         this.level.getProfiler().pop();
++        // Paper start - track when all ticks are required for the entity, e.g. after a piston push
++        if (this.isTickRequired()) {
++           this.remainingExactTicks -= 1;
```
Personally, for counters I like using `--` instead of ` -= `, I think it just signals this is counting smth easily.

petersv5 replied:
Want me to update, or will you do that during merge?

petersv5 replied:
Updated as requested.

Proximyst reviewed (approved):
@kennytv :dinkdonk:

kennytv commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
+                                 }
+ 
+                                 entity.setDeltaMovement(e, g, h);
++                                entity.requestExactTicks(2); // Paper - track when all ticks are required for the entity, e.g. after a piston push
```
2 is not enough (and use the current tick time +). Also add `this.activatedTick = MinecraftServer.currentTick + <whatever would be appropriate>;`, probably 10-20

petersv5 replied:
Do you want me to switch from tracking/counting down) a certain number of ticks to instead track the last tick to be immune in absolute terms?

The patch originally started with 10 ticks, but Proimyst asked it to be reduced. 2 ticks are sufficient for entities stuck in pistons, as they will complete the movment in 2 ticks from reading the code and testing. The original 10 ticks were to allow for entities being bounced (not stuck inside slime) to complete ballisitc movement. After some back and forth we decided to drop that as we had no test cases for how long was sufficient anyway.  Do you want me to up the immunity again to 20 ticks (one second)?

Whis is the semantic for the this.activatedTick?

kennytv replied:
If you are 100% certain 2 ticks are enough, then that's fine. In my testing it wasn't but I had some other code changes flying around as well, so idk. activatedTick is basically exactly what you are doing without the force part, since it additionally does that skip every 4 ticks if not immune. so if activatedTick >= currentTick, the entity should be ticked (so you do the same as that, but checked in the immunity method).

I don't have the code open anymore, but if the immunity check method does not have any other usages, you can keep your check at the place it is currently at to return true as being active. Not sure if you can skip setting activatedTick tho, but it doesn't hurt eitherway

petersv5 replied:
I may not follow exactly here. Did you mean to both move activatedTime approriately into the future + also having an entity-centric immunization as in my patch but with updated semantics?

kennytv replied:
Yesn't to the first, since I wasn't sure. Looking at the activatedTick usages, you don't necessarily have to add the codeline I provided. Just keep the number at whatever you are sure works and only update the immunity `entity.activationImmunityTick = MinecraftServer.currentTick + <ticks it should be active for>;` (your countdown variable but made into absolute tick time)

petersv5 replied:
Yes, my understanding is that the entity is already activated in this particular case and that it is just the decimation-by-4 that is casuing the intermittent issues.

One thing strikes me: would it make more sense to say that activatedTick in the future should be immune to the decimation? I have not followed all the other uses of the activatedTick, maybe that would completely neuter the load improvement from the decimation?

kennytv replied:
If activatedTick is behind the currentTick, the entity is marked as inactive and doesn't tick at all unless it is immune given by the immune check in ActivationRange. Sidenote: If *this* particular spot didn't exist and you only had to account for the other place in the move() method, you could even just use `entity.pistonDeltasGameTime` and shove it into the immunity method.. but you can't, since this is here

So the correct way of keeping it active is by "immunizing" it - activatedTick active entities are supposed to be skipped every 4th unless "immune". But since the immunity check method is used nowhere else, it doesn't matter wheter your check is put into that ActivationRange method or exactly where you have it at the moment (tho you *do* have to set activatedTick here *if* you moved it into the immune check method).

petersv5 replied:
A little more digging on my part: Today the activation of entities moved by being stuck on honey or bounced _on_ slime blocks (and some other cases) get activated fo 20 ticks in Entity.move() due to the movementType being MoverType.PISTON. For items that are stuck _in_ slime instead get moved by Entity.setDeltaMovement that keeps being updated by the PistonMovingBlockEntity. 

In my patch I did the immunization for both cases in PistonMovingBlockEntity. However, maybe the immunzation for the non-stuck-in-slime should be moved next to the activatedTime update in Entity.move() ?

What concerns me is that I cannot find where the entities stuck inside slime received their activation before my patch. They were clearly activated as the failure rate was 1/4. Was it due to short distance to the player? I propose to add setting both activatedTick and actiavtedImmunityTick when doing the setDeltaMovememnt call and moving the test into the checkEntityimmunities() as you suggested.

kennytv replied:
Yep, that should work. 👍

petersv5 replied:
Is it ok to blindly assign activatedTick / activatedimmunityTick with currentTick+xxxx, or sohuld I use max(activatedtick, currentTick+xx) ?

kennytv replied:
Good point. The piston type in the move method currently is the only one that would theoretically not recover itself, but better save then sorry

petersv5 replied:
MinecraftServer does not seem to be used in PistonMovingBlockEntity at the moment. What is the best way to get at the currentTick there? Should I just import it, or is there some other way?

kennytv replied:
I believe it is a static value, so using a fully qualified import, `net.minecraft.server.MinecraftServer.currentTick`

kennytv commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
+     }
++ 
++    // Track when all ticks are required for the entity, e.g. after a piston push
++    private int remainingExactTicks = 0;
```
Make this public, move it under `activatedTick`, store a ticktime, remove getter/setter, and preferably rename this to something like `activationImmunityTick`.

petersv5 replied:
Oki. I was asked to make it private (was origianlly public) and introduce getters/setters. Are you sure you want this to be direct access instead of setter/getter?

kennytv replied:
Yes, to make it more in-line with the original EAR patch once merged into it.

kennytv commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
+     private static void moveEntityByPiston(Direction direction, Entity entity, double d, Direction direction2) {
+         NOCLIP.set(direction);
+         entity.move(MoverType.PISTON, new Vec3(d * (double)direction2.getStepX(), d * (double)direction2.getStepY(), d * (double)direction2.getStepZ()));
++        entity.requestExactTicks(10); // Paper - track when all ticks are required for the entity, e.g. after a piston push
```
Move this into the move method where `this.activatedTick` is assigned (and use the current tick time + 20)

kennytv commented on patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch:
```diff
@@ -0,0 +1,80 @@
+         if (entity instanceof Mob && ((Mob)entity).leashHolder instanceof Player) {
+             return true;
+         }
++        // immunize entities requesting exact ticks, e.g. being affected by a piston
```
Directly check against the public immunity tick time ~~and move the check into `ActivationRange#checkEntityImmunities`~~

kennytv reviewed (changes requested):
Put this into the EAR patch (0364) if you can please. Most of the requested changes are for a cleaner diff. And please make sure if properly works after that one more time, I only briefly tested the ice machine linked in the issue.

kennytv commented on patches/server/0364-Entity-Activation-Range-2.0.patch:
```diff
@@ -333,8 +335,31 @@ index af0dd46e352115d1984f07fff746c7ad41486ac2..fb6da4bccd33b9e079f4bcf84f8ff59a
+ 
+                                 entity.setDeltaMovement(e, g, h);
++                                // Paper - EAR items stuck in in slime pushed by a piston
++                                entity.activatedTick = Math.max(entity.activatedTick, MinecraftServer.currentTick + 10); 
```
Do a fully qualified import here and remove the class import

kennytv reviewed (changes requested):
Don't worry about the failed build, that has nothing do with the pr. Rebasing would be nice - not sure if it already fixes the build, but if not, the merger will rebase again.

Commit:
Fix entities glitching through blocks pushed by pistons (#5189)

Fix #5189 where entities in slime blocks are unreliable.

Add a general mechanism where the entity knows how many ticks into
the future that it has to be fully ticked. ActivationRange respects
that and does not cull any of those ticks to the entity.

kennytv reviewed (approved):
Thank you! We'll see about merging once the build issues are fixed
